Feature/ban

DokyungYou opened the pull request:
### 변경사항
**AS-IS**
신고기능 수정, 서비스 테스트코드, 실행 테스트

**TO-BE**
컨트롤러 통합 테스트 수정

### 테스트
- [x] 테스트 코드
- [x] API 테스트 


Refactor: 신고기능 수정 - [b67f589] 부터입니다!

DokyungYou committed:
feature: 신고(+ 밴 및 영구정지) 기능 구현

DokyungYou committed:
Refactor: 신고기능 수정

- 신고 관련 엔티티, 코드 구조 변경
- native 쿼리 사용한 메소드 -> 쿼리 dsl 로 변경

DokyungYou committed:
Refactor,Test: 일부 코드 수정,신고기능 서비스 단위테스트 ,

- MemberServiceTest3
- @AuthenticationPrincipal Member -> MemberDetail
- native 쿼리 사용한 메소드 -> 쿼리 dsl 로 변경
- 기존의 존재하지 않는 회원조건에 탈퇴회원 조건 추가

DokyungYou committed:
Refactor,Test: 일부 코드 수정,신고기능 컨트롤러 통합테스트

DokyungYou committed:
Merge branch 'develop' into feature/ban

# Conflicts:
#	solumon-backend/build.gradle
#	solumon-backend/src/main/java/com/example/solumonbackend/global/exception/ErrorCode.java
#	solumon-backend/src/main/java/com/example/solumonbackend/member/entity/Member.java
#	solumon-backend/src/main/java/com/example/solumonbackend/member/repository/MemberRepository.java
#	solumon-backend/src/main/java/com/example/solumonbackend/member/service/MemberService.java

ofgongmu commented on solumon-backend/src/test/java/com/example/solumonbackend/member/service/MemberServiceTest3.java:
```diff
@@ -0,0 +1,274 @@
+
+
+  private Member fakeMember1;
+  private MemberBuilder otherMemberBuilder;
```
Member이랑 MemberBuilder 둘 중에 하나만 써도 될 것 같은데 혹시 둘을 분리하신 이유가 있을까요? @beforeeach에서도요!

DokyungYou replied:
기준이 되는 멤버 외에  
탈퇴날짜만 있는 객체,
정지날짜만 있는 객체,
두 날짜 값이 없는 객체가 필요한 상황이어서
객체를 세 개 만들어 놓는 것 보다 
otherMember는 빌더로 빼놓고 상황에 맞게 값을 할당해서 쓰자는 생각으로 만들었었어요.

기준이 되는 멤버는 모든 메소드에 그대로 쓰이는 부분이 많아서 메소드 안에서 전역변수를 바로 넣어주는 방식으로 사용하려했습니다

DokyungYou replied:
다른 분 리뷰로 멤버 메소드에  객체가 아닌 아이디값을 넣는걸로 변경했기때문에
현재는
해당 테스트 코드의 기준 Member는 지우고 Lond memberId만,
otherMember는 빌더 그대로 남겼습니다

ofgongmu reviewed (commented):
신고가 생각보다 복잡한 기능이었네요 고생 많으셨습니다!

DokyungYou committed:
Test: 컨트롤러 테스트 코드 수정

- MemberControllerTest3 수정
- 현재 브랜치의 엔티티 구조가 바뀐 문제때문에 실행이 안돼서
develop에서 pull해온 다른 팀원 테스트 코드에서 .reportCount(0) 주석 처리

Gayeon-H commented on solumon-backend/src/main/java/com/example/solumonbackend/member/entity/Report.java:
```diff
@@ -0,0 +1,42 @@
+  private Long reportId;
+
+  @ManyToOne
+  private Member member;
```
여기 JoinColumns("member_id") 어노테이션 빼먹으신 것 같아요

Gayeon-H commented on solumon-backend/src/main/java/com/example/solumonbackend/member/repository/MemberRepository.java:
```diff
@@ -1,13 +1,17 @@
 
   Optional<Member> findByNickname(String nickName);
+
+  Optional<Member> findByMemberId(Long id);
```
findById가 있는데 따로 findByMemberId를 만드신 이유가 있나요?

DokyungYou replied:
pk 컬럼의 이름이 memberId여서 저렇게 만들어서 사용했는데
혹시 말씀하시는 findById가 
  @Override
  Optional<Member> findById(Long aLong); 이 것 말씀하시는 건가요???

DokyungYou replied:
JpaRepository에 이게 있다는 걸 생각치도 못하고 매번 직접 만들어서 쓰고 있었네요.

Gayeon-H commented on solumon-backend/src/main/java/com/example/solumonbackend/member/repository/ReportRepository.java:
```diff
@@ -0,0 +1,20 @@
+public interface ReportRepository extends JpaRepository<Report, Long> {
+
+
+  Optional<Report> findTopByMemberAndReporterIdOrderByReportedAtDesc(Member member, Long reporterId);
```
저번에 받은 멘토님 피드백처럼 member 대신 memberId로 수정하셔야 할 것 같아요!

DokyungYou replied:
저번에 그렇게 수정해놓고 잊고있었네요.
상기시켜주셔서 감사합니다

Gayeon-H commented on solumon-backend/src/main/java/com/example/solumonbackend/member/service/MemberService.java:
```diff
@@ -94,4 +100,77 @@ public GeneralSignInDto.Response signIn(GeneralSignInDto.Request request) {
+    if(optionalMember.isEmpty() || optionalMember.get().getUnregisteredAt() != null ){
+      throw new MemberException(NOT_FOUND_MEMBER);
+    }
+    Member reportedMember = optionalMember.get();
```
제가 저번에 optional에서 get 썼다가 쓰지 말라는 피드백을 받아서 orElseThrow로 1차적으로 null 걸러준 다음에 unregisterdAt이 null인지 확인해주는 건 어떨까요?

DokyungYou replied:
비어있거나 회원탈퇴 날짜가 있는 경우 둘 다 같은 NOT_FOUND_MEMBER예외를 던지게 해서 한번에 묶느라고 그렇게 썻었어요.
말씀하신대로 orElseThrow로 바꾸고 reportedMember.getUnregisteredAt() != null의 경우 UNREGISTERED_MEMBER에러코드 사용하는 걸로 수정했습니다!

Gayeon-H reviewed (commented):
리뷰 드렸습니다. 수고 많으셨어요!!

DokyungYou committed:
Rename: 테스트클래스 이름 변경

DokyungYou committed:
Refactor: 리뷰 반영하여 수정

DokyungYou committed:
Refactor:  JpaRepository 의 findById로 사용하는것으로 변경

- MemberRepository 에서 findByMemberId로 만들어서 사용했던 것을 삭제

ofgongmu reviewed (approved):
확실히 리뷰 이후로 수정해주신 부분들이 전보다 알아보기 좋은 것 같아요. 고생하셨습니다!

Gayeon-H reviewed (approved):
수정하신다고 고생 많으셨어요!

DokyungYou committed:
Chore: develop pull 해온 데이터와 merge

- develop에 merge 전 충돌되는 코드들 해결

DokyungYou commented:
Feature/ban Merge to develop
